src/search_controller.ts

```diff
@@ -0,0 +1,484 @@
+  }
+
+  get isActive() {
+    return this.state.getLatestValue().isActive;
+  }
+
+  get isLoading() {
+    return this.state.getLatestValue().isLoading;
+  }
+
+  get items() {
+    return this.state.getLatestValue().items;
+  }
+
+  get next() {
+    return this.state.getLatestValue().next;
+  }
+
+  get offset() {
+    return this.state.getLatestValue().offset;
+  }
+
+  get searchQuery() {
+    return this.state.getLatestValue().searchQuery;
+  }
+
+  protected abstract query(searchQuery: string): Promise<QueryReturnValue<T>>;
+
+  protected abstract filterQueryResults(items: T[]): T[] | Promise<T[]>;
+
+  setDebounceOptions = ({ debounceMs }: DebounceOptions) => {
+    this.searchDebounced = debounce(this.executeQuery.bind(this), debounceMs);
+  };
+
+  activate = (sourceStateOverride?: Partial<SearchSourceState>) => {
+    if (this.isActive) return;
+    if (this.searchQuery) {
+      this.search();
+    }
+    this.state.partialNext({ ...sourceStateOverride, isActive: true });
+  };
+
+  deactivate = () => {
+    if (!this.isActive) return;
+    this.state.partialNext({ isActive: false });
+  };
+
+  async executeQuery(newSearchString?: string) {
+    const hasNewSearchQuery = typeof newSearchString !== 'undefined';
+    const searchString = newSearchString ?? this.searchQuery;
+    if (!this.isActive || this.isLoading || (!this.hasMore && !hasNewSearchQuery) || !searchString) return;
+
+    if (hasNewSearchQuery) {
+      this.resetState({ isActive: this.isActive, isLoading: true, searchQuery: newSearchString });
+    } else {
+      this.state.partialNext({ isLoading: true });
+    }
+
+    const stateUpdate: Partial<SearchSourceState<T>> = {};
+    try {
+      const results = await this.query(searchString);
+      if (!results) return;
+      const { items, next } = results;
+
+      if (next) {
+        stateUpdate.next = next;
+        stateUpdate.hasMore = !!next;
+      } else {
+        stateUpdate.offset = (this.offset ?? 0) + items.length;
+        stateUpdate.hasMore = items.length === this.pageSize;
+      }
+
+      stateUpdate.items = await this.filterQueryResults(items);
+    } catch (e) {
+      stateUpdate.lastQueryError = e as Error;
+    } finally {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      this.state.next(({ lastQueryError, ...prev }: SearchSourceState<T>) => ({
+        ...prev,
+        ...stateUpdate,
+        isLoading: false,
+        items: [...(prev.items ?? []), ...(stateUpdate.items || [])],
+      }));
+    }
+  }
+
+  search = (searchQuery?: string) => {
+    this.searchDebounced(searchQuery);
+  };
+
+  resetState(stateOverrides?: Partial<SearchSourceState<T>>) {
+    this.state.next({
+      hasMore: true,
+      isActive: false,
+      isLoading: false,
+      items: undefined,
+      lastQueryError: undefined,
+      next: undefined,
+      offset: 0,
+      searchQuery: '',
+      ...stateOverrides,
+    });
```

Comment:
Where `initialState` would get set at the beginning (during construction) and would be `private readonly` property.

```suggestion
    this.state.next(this.initialState)
```